### ISSUE TITLE
`oximeter_db` could use a more efficient serialization protocol

### ISSUE THREAD

**Issue by bnaecker** [oxidecomputer/omicron#4059]
There's a huge amount of repetitive code in `oximeter_db::model` to handle serialization. We're using the HTTP interface and the `JSONEachRow` format, which is easy but inefficient and verbose. The [`RowBinary`](https://clickhouse.com/docs/en/interfaces/formats#rowbinarywithnamesandtypes) and related, and even the [`Native`](https://clickhouse.com/docs/en/interfaces/formats#native) formats, are very straightforward. They should also be dramatically more efficient, even when using the HTTP interface. We should consider moving to these.

**Comment by bnaecker** [comment]
There are other reasons than performance to prioritize this change. We support a number of floating point metrics, the simplest being `f64` and (soon) `f32` gauge types. Since we're using `serde_json` as the serialization layer with the database, we'll pass an `f32` through `serde_json::Number`. Looking at the code, we see this:

https://github.com/serde-rs/json/blob/b6e113f2036c52e994ca805e530ee4ffae791f71/src/number.rs#L320-L336

So that takes any `f32` and casts it directly to an `f64`. That's not awesome. Since doubles have higher precision, we'll actually send the wrong value into the database. Assuming it's sent to the right table (with an `Float32` datatype), it will be truncated back, so the right value will be _stored_ in the table. But we're still serializing the wrong thing.

On  related point, if the values happen to have a nice integer representation, say `1.0`, then ClickHouse sends back JSON that interprets the value as an _integer_ rather than a float. For example:

```
shale :) select 1.0 as foo format JSONEachRow;

SELECT 1. AS foo
FORMAT JSONEachRow

Query id: edd2b0ff-461c-4525-b80e-641aa76350b8

{"foo":1}

1 rows in set. Elapsed: 0.002 sec.
```

This will cause `serde_json` to interpret that values as an integer, not a float, e.g., `Number(1)`, rather than `Number(1.0)`.

To be clear, as long as we convert the resulting `Value` into the right type, e.g., by calling `.as_f32()`, after fetching it from the database, we should round-trip the expected values. We'll not lose any precision here, this is just a really confusing wart that arises from using JSON and how ClickHouse reports values back. We could obviate all this by using one of the binary protocols listed in the issue.

**Comment by bnaecker** [comment]
For completeness, [here's](https://clickhouse.com/docs/en/interfaces/formats#tabseparated-data-formatting) the ClickHouse documentation around floating-point formatting. This is particularly annoying:

> During formatting, accuracy may be lost on floating-point numbers.

**Comment by bnaecker** [comment]
This will be closed by #6584
